core/src/main/java/org/mule/runtime/core/internal/exception/GlobalErrorHandler.java

```diff
@@ -7,26 +7,95 @@
 package org.mule.runtime.core.internal.exception;
 
 import org.mule.runtime.api.component.location.Location;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lifecycle.InitialisationException;
 import org.mule.runtime.core.api.event.CoreEvent;
-import org.mule.runtime.core.privileged.exception.MessagingExceptionHandlerAcceptor;
 import org.mule.runtime.core.privileged.exception.TemplateOnErrorHandler;
+
 import org.reactivestreams.Publisher;
 
-import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 
-import static java.util.stream.Collectors.toList;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.System.getProperty;
+import static org.mule.runtime.api.util.MuleSystemProperties.REUSE_GLOBAL_ERROR_HANDLER_PROPERTY;
 
 public class GlobalErrorHandler extends ErrorHandler {
 
+  private final AtomicBoolean initialised = new AtomicBoolean(false);
+  private final AtomicInteger started = new AtomicInteger(0);
+  private final AtomicBoolean disposed = new AtomicBoolean(false);
+
   @Override
   public Publisher<CoreEvent> apply(Exception exception) {
     throw new IllegalStateException("GlobalErrorHandlers should be used only as template for local ErrorHandlers");
   }
 
+  @Override
+  public void initialise() throws InitialisationException {
+    if (!reuseGlobalErrorHandler()) {
+      super.initialise();
+      return;
+    }
+
+    if (!initialised.getAndSet(true)) {
+      setFromGlobalErrorHandler();
```

Comment:
this should be set during construction, not in initialization. Maybe `ErrorHandlerFactoryBean` is a better place to do this